### ISSUE TITLE
Support FastAPI dependency overrides

### ISSUE THREAD

**Issue by rowanbarrie** [meadsteve/lagom#174]
Hi Steve,

Love the framework so far. One issue I'm having is in trying to use [FastAPI dependency overrides](https://fastapi.tiangolo.com/advanced/testing-dependencies/) in a similar manner to vanilla FastAPI, when making web-layer unit tests. I.e. mocking out dependencies.

In vanilla FastAPI you would do something like this (with pytest):

```
...
from fastapi.testclient import TestClient
from dependencies import get_foo_service  # <-- Dependency function
from foo import FooService
...

@pytest.fixture
def foo_service_mock():
    return Mock(FooService)

@pytest.fixture
def client(foo_service_mock) -> Generator:
    app.dependency_overrides[get_foo_service] = lambda: foo_service_mock

    with TestClient(app) as c:
        yield c
```

But I can't see how I could replicate the override (`app.dependency_overrides[get_foo_service]`) functionality when using the Lagom FastAPI DI approach (i.e. when using `deps.depends(FooService)`). 

I suppose I would somehow need to reference the [inner `_resolver` function ](https://github.com/meadsteve/lagom/blob/39b5ce620534fb3e77f2a290cfc2e8f3a50d4005/lagom/integrations/fast_api.py#L41) of `depends`...surely this isn't possible?

Thus it would be nice if there were explicit support for FastAPI dependency overrides.

Thanks!

**Comment by meadsteve** [comment]
@rowanbarrie thanks for raising this. I can't think of a clean way of doing this with Lagom+FastAPI currently. I'll think a bit more about this and see if there's a sensible equivalent of the fastpi overrides that lagom could provide.

**Comment by meadsteve** [comment]
Lagom already has a concept of building up layers of overrides being built up for containers so it should be possible to use this. I'm thinking something like this:

```python
...
from fastapi.testclient import TestClient
from foo import FooService
from your_app import app, deps
...

@pytest.fixture
def foo_service_mock():
    return Mock(FooService)

@pytest.fixture
def client(foo_service_mock) -> Generator:
   
    with TestClient(app) as c, deps.override_container_for_test(container: container[FooService] = foo_service_mock ):
        yield c
```

I'm imaging the ´override_container_for_test` method would take a function which updates the dependency container with any changes needed for test and returns a context manager which resets deps on exit.

**Comment by meadsteve** [comment]
I went for a slightly different approach in #177 and returned a mutable container from the context manager rather than providing an update function/lambda. It felt easier to use and read.

**Comment by meadsteve** [comment]
The work in #177 is now released as version 1.6.0. Let me know how you get on @rowanbarrie . I'll close this issue but we can re-open it if there's some use-cases I've not covered.

**Comment by rowanbarrie** [comment]
Works a charm, thanks @meadsteve! 🍻
